### PR TITLE
[FIX] account: general performances

### PR DESCRIPTION
**Pull request by william-andre**
* The default order of `account.move.line` includes the `sequence` since 080d41a34119191d20b3e07f72f6aeec677ee1a8 
  This field was not indexed and was leading to poor performances when opening the list view of journal items or when searching without overriding the order.
* A typo in `_get_invalid_statement_ids` was leading to recomputing all the validity of all statements when we only needed to compute a few
* Domains on `account.move.line` often include filters on related fields, including the accounts' and payments' types. This can lead to multiple sub queries that are repeating the same thing.

### CODE REVIEW

File: addons/account/models/account_bank_statement.py [hunk 1]
```diff
@@ -223,20 +223,24 @@ def create(self, vals_list):
...
+                   LATERAL (
+                       SELECT balance_end_real
+                         FROM account_bank_statement st_lookup
+                        WHERE st_lookup.first_line_index < st.first_line_index
+                          AND st_lookup.journal_id = st.journal_id
+                     ORDER BY st_lookup.first_line_index desc
+                        LIMIT 1
+                   ) prev
+             WHERE prev.balance_end_real != st.balance_start
+               {"" if all_statements else "AND st.id IN %(ids)s"}
```
Comment (jco-odoo) [inline addons/account/models/account_bank_statement.py:241]: seems like a regular bug fix (just inversion of all_statements logic)

File: addons/account/models/account_move_line.py [hunk 2]
```diff
@@ -17,7 +17,7 @@ class AccountMoveLine(models.Model):
     _name = "account.move.line"
     _inherit = "analytic.mixin"
     _description = "Journal Item"
-    _order = "date desc, move_name desc, sequence, id"
+    _order = "date desc, move_name desc, id"
```
Comment (jco-odoo) [inline addons/account/models/account_move_line.py:20]: but is not this one important on invoice lines?

File: addons/account/models/account_move_line.py [hunk 3]
```diff
@@ -78,6 +78,7 @@ class AccountMoveLine(models.Model):
         compute='_compute_account_id', store=True, readonly=False, precompute=True,
         inverse='_inverse_account_id',
         index=True,
+        auto_join=True,
```
Comment (jco-odoo) [inline addons/account/models/account_move_line.py:81]: so no security issue then?

File: addons/l10n_account_edi_ubl_cii_tests/tests/common.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 import base64
 
 from freezegun import freeze_time
+from collections import Counter
```
Comment (jco-odoo) [inline addons/l10n_account_edi_ubl_cii_tests/tests/common.py:5]: unnecessary diff

Comment (william-andre) [reply addons/l10n_account_edi_ubl_cii_tests/tests/common.py:5]: What? No.

This is because the default order on invoice lines changed (removed the sequence), so the assertion values are not in the same order anymore.
But we don't care about the order so I just used a counter.

File: addons/account/models/account_move_line.py (hunk 3 above)
Comment (william-andre) [reply addons/account/models/account_move_line.py:81]: I think that it is fine. If you have access to `account.move.line`, you should have access to those auto joined fields

File: addons/account/models/account_move_line.py (hunk 2 above)
Comment (william-andre) [reply addons/account/models/account_move_line.py:20]: No, the order is explicitly `sequence` in the view. I don't think that it has a meaningful impact anywhere else.

File: addons/account/models/account_bank_statement.py (hunk 1 above)
Comment (william-andre) [reply addons/account/models/account_bank_statement.py:241]: It isn't a "bug" as it only impacts the perfs; the result is still correct. (except for the search function, but...)

**Comment by jco-odoo** [comment]
@robodoo r+
